Content.Server/Automod/AutomodManager.cs

```diff
@@ -0,0 +1,260 @@
+
+    private ISawmill _log = default!;
+
+    // Filters
+    private readonly Dictionary<AutomodTarget, Dictionary<Regex, AutomodFilterDef>> _regexFilters = new();
+
+    #region Initialize
+
+    public void Initialize()
+    {
+        IoCManager.InjectDependencies(this);
+
+        LoadAutomodFiltersFromDatabase();
+    }
+
+    private async void LoadAutomodFiltersFromDatabase()
+    {
+        var filters = await _db.GetAllAutomodFiltersAsync();
+        foreach (var filter in filters)
+        {
+            AddFilter(filter);
+        }
+    }
+
+    void IPostInjectInit.PostInject()
+    {
+        _log = _logMan.GetSawmill("automod");
+    }
+
+    #endregion
+
+    #region Filter
+
+    public bool Filter(AutomodTarget target, string inputText, ICommonSession session)
+    {
+        return RegexFilter(target, inputText, session);
+    }
+
+    private bool RegexFilter(AutomodTarget target, string inputText, ICommonSession session)
+    {
+        // Ensure that only 1 bit is set
+        Debug.Assert(target != 0 && (target & (target - 1)) == 0);
+
+        var blocked = true;
+
+        // No filters defined for target
+        if (!_regexFilters.TryGetValue(target, out var automodFilters))
+            return true;
+
+        foreach (var (regex, filter) in automodFilters)
+        {
+            var regexMatches = regex.Matches(inputText);
+            if (regexMatches.Count == 0)
+                continue;
+
+            var textMatches = new Dictionary<string, int>();
+            foreach (Match match in regexMatches)
+            {
+                var str = match.ToString();
+                textMatches.Add(str, match.Index);
+            }
+
+            if (!_protoMan.TryIndex<AutomodActionGroupPrototype>(filter.ActionGroup, out var censorGroup))
+            {
+                _log.Error($"AutomodActionGroupPrototype \"{filter.ActionGroup}\" not found.");
+                continue;
+            }
+
+            var skip = false;
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                if (!censorAction.Skip(inputText, textMatches))
+                    continue;
+
+                skip = true;
+                break;
+            }
+            if (skip)
+                continue;
+
+            var displayName = GetDisplayName(filter);
+
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                blocked &= censorAction.RunAction(session, inputText, textMatches, filter, displayName, _entMan);
+            }
+        }
+
+        return blocked;
+    }
+
+    private string GetDisplayName(AutomodFilterDef filter)
+    {
+        return filter.Id != null ? $"{filter.DisplayName} (automod#{filter.Id})" : $"{filter.DisplayName} (automod#na)";
+    }
+
+    #endregion
+
+    #region Reload
+
+    public async void ReloadAutomodFilters()
+    {
+        _regexFilters.Clear();
+
+        var filters = await _db.GetAllAutomodFiltersAsync();
```

Comment:
This duplicates the initialize loading code.

Reply:
Removed [ec876a7](https://github.com/space-wizards/space-station-14/pull/28570/commits/ec876a7919e96fe79c37bfc735ead3f0e40befbe)